presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -252,6 +260,96 @@ public HiveTableHandle getTableHandle(ConnectorSession session, SchemaTableName
         return new HiveTableHandle(tableName.getSchemaName(), tableName.getTableName());
     }
 
+    private static final String PARTITIONS_TABLE_SUFFIX = "$partitions";
+
+    @Override
+    public Optional<SystemTable> getSystemTable(ConnectorSession session, SchemaTableName tableName)
+    {
+        if (tableName.getTableName().endsWith(PARTITIONS_TABLE_SUFFIX)) {
+            return getPartitionsSystemTable(session, tableName);
+        }
+        return Optional.empty();
+    }
+
+    private Optional<SystemTable> getPartitionsSystemTable(ConnectorSession session, SchemaTableName tableName)
+    {
+        SchemaTableName sourceTableName = new SchemaTableName(
+                tableName.getSchemaName(),
+                tableName.getTableName().substring(0, tableName.getTableName().length() - PARTITIONS_TABLE_SUFFIX.length()));
```

Comment:
What if `tableName.getTableName().substring(0, tableName.getTableName().length() - PARTITIONS_TABLE_SUFFIX.length())` is `""` ?